From d3cd3554d44159d63b72173aaff74be67fc52131 Mon Sep 17 00:00:00 2001
From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Wed, 12 Feb 2025 13:08:43 +0800
Subject: [PATCH 1/2] mtk: mt76: fix kernel warning for of_node_put during
 module reload

Remove the unnecessary of_node_put in mt76_check_bin_file_mode;
otherwise, it will trigger the following kernel warning.
To be more specific, of_node_put decrements the refcount of a node,
and it is being used to release a node obtained by of_node_get
(e.g. of_get_parent, of_find_node_by_xxx, etc).
So, of_node_put is not necessary after calling of_property_read_string since
the function only finds and reads the property in the current node.
Kernel warning:
OF: ERROR: of_node_release() detected bad of_node_put() on /soc/pcie@11300000/pcie@0,0/mt7996@0,0
CPU: 1 PID: 10618 Comm: insmod Tainted: G           O       6.6.73 #0
Hardware name: MediaTek MT7988A Reference Board (DT)
Call trace:
 dump_backtrace+0x9c/0xd8
 show_stack+0x14/0x1c
 dump_stack_lvl+0x44/0x58
 dump_stack+0x14/0x1c
 of_node_release+0x120/0x18c
 kobject_put+0xa4/0xf0
 of_node_put+0x14/0x20
 mt76_check_bin_file_mode+0x84/0xa0 [mt76]
 mt7996_eeprom_check_fw_mode+0x28/0x164 [mt7996e]
 mt7996_register_device+0x31c/0xba4 [mt7996e]
 mt7996_pci_probe+0x39c/0x64c [mt7996e]
 pci_device_probe+0x94/0x128
 really_probe+0x140/0x2a8
 _driver_probe_device+0x74/0x120
 driver_probe_device+0x3c/0xd4
 __driver_attach+0x88/0x150
 bus_for_each_dev+0x60/0x9c
 driver_attach+0x20/0x28
 bus_add_driver+0xdc/0x1fc
 driver_register+0x58/0x114
 __pci_register_driver+0x48/0x50
 mt7996_init+0x68/0x1000 [mt7996e]
 do_one_initcall+0x40/0x24c
 do_init_module+0x54/0x1dc
 load_module+0x1a78/0x1b64
 __do_sys_init_module+0x210/0x258
 __arm64_sys_init_module+0x18/0x20
 invoke_syscall.constprop.0+0x4c/0xe0
 do_el0_svc+0x3c/0xbc
 el0_svc+0x18/0x4c
 el0t_64_sync_handler+0x118/0x124
 el0t_64_sync+0x150/0x154
OF: ERROR: next of_node_put() on this node will result in a kobject warning 'refcount_t: underflow; use-after-free.'

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 eeprom.c | 2 --
 1 file changed, 2 deletions(-)

diff --git a/eeprom.c b/eeprom.c
index 6ff22319..0a23bc92 100644
--- a/eeprom.c
+++ b/eeprom.c
@@ -228,8 +228,6 @@ bool mt76_check_bin_file_mode(struct mt76_dev *dev)
 #endif
 	}
 
-	of_node_put(np);
-
 	return dev->bin_file_name ? true : false;
 }
 EXPORT_SYMBOL_GPL(mt76_check_bin_file_mode);
-- 
2.45.2

